File: samples/fb-cpp/schemas.cpp
/*
 *  schemas.cpp (fb-cpp) — name resolution watched through typed statements.
 *
 *  The fb-cpp twin of ../cpp/schemas.cpp: the same five demonstrations —
 *  RDB$SCHEMAS and the default path, one unqualified SELECT resolving
 *  differently as SET SEARCH_PATH changes, SYSTEM auto-appended, stored
 *  code binding its OWN schema forever, and the schema-qualified plan.
 *  Nothing schema-specific needs the API: search path is session state the
 *  engine keeps, so the twin is a straight port — the plan simply comes
 *  from StatementOptions().setPrefetchLegacyPlan() instead of a getPlan()
 *  call on IStatement.  See ../../schemas-and-name-resolution.md.
 *
 *  Build & run (see ../README.md):
 *      ./build/fbcpp_schemas [database]
 */

#include "fbcpp_sample.h"
#include <cstdio>
#include <string>

using namespace fbcpp;
using namespace fbcpp_sample;

static void exec(Attachment& att, Transaction& tra, const char* sql)
{
	Statement{att, tra, sql}.execute(tra);
}

static std::string one(Attachment& att, Transaction& tra, const char* sql)
{
	Statement stmt{att, tra, sql};
	stmt.execute(tra);
	return stmt.getString(0).value_or("<null>");
}

static std::string ctxPath(Attachment& att, Transaction& tra)
{
	return one(att, tra,
		"SELECT RDB$GET_CONTEXT('SYSTEM','SEARCH_PATH') FROM RDB$DATABASE");
}

int main(int argc, char** argv)
{
	const char* database = argOrDefault(argc, argv, 1,
		"inet://localhost//tmp/fbhandson/schemas_fbcpp.fdb");

	try
	{
		Client client{"fbclient"};
		Attachment att = attachOrCreate(client, database);

		// -- Idempotent cleanup + setup.
		{
			Transaction ddl{att};
			for (const char* s : { "DROP PROCEDURE APP.WHICH_ONE",
					"DROP TABLE PUBLIC.CUSTOMERS", "DROP TABLE APP.CUSTOMERS",
					"DROP SCHEMA APP" })
				try { exec(att, ddl, s); } catch (const DatabaseException&) {}
			exec(att, ddl, "CREATE SCHEMA APP");
			exec(att, ddl, "CREATE TABLE PUBLIC.CUSTOMERS (ID INT, ORIGIN VARCHAR(20))");
			exec(att, ddl, "CREATE TABLE APP.CUSTOMERS    (ID INT, ORIGIN VARCHAR(20))");
			ddl.commit();
		}

		// -- 1. The catalog and the default path.
		Transaction tra{att};
		exec(att, tra, "INSERT INTO PUBLIC.CUSTOMERS VALUES (1, 'from PUBLIC')");
		exec(att, tra, "INSERT INTO APP.CUSTOMERS    VALUES (2, 'from APP')");
		printf("schemas in RDB$SCHEMAS      : ");
		{
			Statement s{att, tra,
				"SELECT TRIM(RDB$SCHEMA_NAME) FROM RDB$SCHEMAS ORDER BY 1"};
			for (bool ok = s.execute(tra); ok; ok = s.fetchNext())
				printf("%s  ", s.getString(0)->c_str());
		}
		printf("\ndefault search path         : %s\n", ctxPath(att, tra).c_str());

		// -- 2. Same statement, two resolutions.
		printf("\nSELECT ORIGIN FROM CUSTOMERS, as the path changes:\n");
		printf("  path PUBLIC,SYSTEM        -> %s\n",
			one(att, tra, "SELECT ORIGIN FROM CUSTOMERS").c_str());
		exec(att, tra, "SET SEARCH_PATH TO APP, PUBLIC");
		printf("  path APP,PUBLIC           -> %s\n",
			one(att, tra, "SELECT ORIGIN FROM CUSTOMERS").c_str());

		// -- 3. SYSTEM can be moved but not removed.
		exec(att, tra, "SET SEARCH_PATH TO APP");
		printf("\nSET SEARCH_PATH TO APP      -> %s   (SYSTEM auto-appended)\n",
			ctxPath(att, tra).c_str());

		// -- 4. Stored code binds its own schema, not the caller's path.
		exec(att, tra, "SET SEARCH_PATH TO APP, PUBLIC");
		exec(att, tra,
			"CREATE PROCEDURE WHICH_ONE RETURNS (SRC VARCHAR(20)) AS "
			"BEGIN SELECT ORIGIN FROM CUSTOMERS INTO :SRC; SUSPEND; END");
		tra.commit();

		Transaction tra2{att};
		printf("\nprocedure created with path APP,PUBLIC (lands in APP, binds APP.CUSTOMERS)\n");
		exec(att, tra2, "SET SEARCH_PATH TO PUBLIC");
		printf("  after SET SEARCH_PATH TO PUBLIC:\n");
		printf("    direct SELECT ... FROM CUSTOMERS -> %s\n",
			one(att, tra2, "SELECT ORIGIN FROM CUSTOMERS").c_str());
		printf("    SELECT SRC FROM APP.WHICH_ONE    -> %s   <- unmoved\n",
			one(att, tra2, "SELECT SRC FROM APP.WHICH_ONE").c_str());
		printf("    RDB$DEPENDENCIES records         -> %s\n",
			one(att, tra2,
				"SELECT TRIM(RDB$DEPENDED_ON_SCHEMA_NAME) || '.' || TRIM(RDB$DEPENDED_ON_NAME)"
				" FROM RDB$DEPENDENCIES WHERE RDB$DEPENDENT_NAME = 'WHICH_ONE'").c_str());

		// -- 5. Plans are schema-qualified.
		{
			Statement s{att, tra2, "SELECT COUNT(*) FROM CUSTOMERS",
				StatementOptions().setPrefetchLegacyPlan(true)};
			printf("\nplan for unqualified SELECT :%s\n", s.getLegacyPlan().c_str());
		}

		tra2.commit();
		printf("\ndone.\n");
		return 0;
	}
	catch (const std::exception& e)
	{
		return report(e);
	}
}

File: samples/fpc/schemas.pas
{
  schemas.pas — schemas and name resolution, replayed through fbintf.

  The fbintf twin of ../cpp/schemas.cpp:

   1. RDB$SCHEMAS and the default search path ("PUBLIC", "SYSTEM").
   2. Two same-named tables (PUBLIC.CUSTOMERS / APP.CUSTOMERS); the same
      unqualified SELECT resolves differently as SET SEARCH_PATH changes.
   3. SYSTEM is auto-appended when omitted from SET SEARCH_PATH.
   4. A stored procedure created while APP leads the path binds
      APP.CUSTOMERS — and keeps meaning that after the session's path
      flips to PUBLIC (stored code resolves in its OWN schema, never the
      caller's path).  RDB$DEPENDENCIES records the resolution.
   5. IStatement.GetPlan shows the schema-qualified plan — fbintf exposes
      the plan directly, no info-buffer decoding needed.

  See ../../schemas-and-name-resolution.md.

  Build & run (see ../README.md):
      make -C samples/fpc bin/schemas && samples/fpc/bin/schemas
}
program schemas;

{$mode delphi}
{$codepage UTF8}
{$interfaces COM}

uses SysUtils, IB, FBHandsOn;

const CleanupDDL: array[0..3] of AnsiString = (
  'DROP PROCEDURE APP.WHICH_ONE',
  'DROP TABLE PUBLIC.CUSTOMERS',
  'DROP TABLE APP.CUSTOMERS',
  'DROP SCHEMA APP');

var A: IAttachment;
    Tr: ITransaction;
    Stmt: IStatement;
    R: IResultSet;
    i: integer;

function NewTr: ITransaction;
begin
  Result := A.StartTransaction([isc_tpb_write, isc_tpb_nowait,
    isc_tpb_read_committed, isc_tpb_rec_version], taCommit);
end;

function Value(const sql: AnsiString): AnsiString;
begin
  Result := A.OpenCursorAtStart(Tr, sql)[0].AsString;
end;

function CtxPath: AnsiString;
begin
  Result := Value('SELECT RDB$GET_CONTEXT(''SYSTEM'',''SEARCH_PATH'') FROM RDB$DATABASE');
end;

begin
  A := AttachOrCreate(DbConn('schemas'));

  { -- Idempotent cleanup + setup. }
  Tr := NewTr;
  for i := low(CleanupDDL) to high(CleanupDDL) do
    try
      A.ExecImmediate(Tr, CleanupDDL[i]);
    except on EIBInterBaseError do ;
    end;
  A.ExecImmediate(Tr, 'CREATE SCHEMA APP');
  A.ExecImmediate(Tr, 'CREATE TABLE PUBLIC.CUSTOMERS (ID INT, ORIGIN VARCHAR(20))');
  A.ExecImmediate(Tr, 'CREATE TABLE APP.CUSTOMERS    (ID INT, ORIGIN VARCHAR(20))');
  Tr.Commit;

  { -- 1. The catalog and the default path. }
  Tr := NewTr;
  A.ExecuteSQL(Tr, 'INSERT INTO PUBLIC.CUSTOMERS VALUES (1, ''from PUBLIC'')', []);
  A.ExecuteSQL(Tr, 'INSERT INTO APP.CUSTOMERS    VALUES (2, ''from APP'')', []);
  write('schemas in RDB$SCHEMAS      : ');
  R := A.OpenCursor(Tr, 'SELECT TRIM(RDB$SCHEMA_NAME) FROM RDB$SCHEMAS ORDER BY 1');
  while R.FetchNext do
    write(R[0].AsString, '  ');
  writeln;
  writeln('default search path         : ', CtxPath);

  { -- 2. Same statement, two resolutions. }
  writeln;
  writeln('SELECT ORIGIN FROM CUSTOMERS, as the path changes:');
  writeln('  path PUBLIC,SYSTEM        -> ', Value('SELECT ORIGIN FROM CUSTOMERS'));
  A.ExecuteSQL(Tr, 'SET SEARCH_PATH TO APP, PUBLIC', []);
  writeln('  path APP,PUBLIC           -> ', Value('SELECT ORIGIN FROM CUSTOMERS'));

  { -- 3. SYSTEM can be moved but not removed. }
  A.ExecuteSQL(Tr, 'SET SEARCH_PATH TO APP', []);
  writeln;
  writeln('SET SEARCH_PATH TO APP      -> ', CtxPath, '   (SYSTEM auto-appended)');

  { -- 4. Stored code binds its own schema, not the caller's path. }
  A.ExecuteSQL(Tr, 'SET SEARCH_PATH TO APP, PUBLIC', []);
  A.ExecImmediate(Tr,
    'CREATE PROCEDURE WHICH_ONE RETURNS (SRC VARCHAR(20)) AS ' +
    'BEGIN SELECT ORIGIN FROM CUSTOMERS INTO :SRC; SUSPEND; END');
  Tr.Commit;

  Tr := NewTr;
  writeln;
  writeln('procedure created with path APP,PUBLIC (lands in APP, binds APP.CUSTOMERS)');
  A.ExecuteSQL(Tr, 'SET SEARCH_PATH TO PUBLIC', []);
  writeln('  after SET SEARCH_PATH TO PUBLIC:');
  writeln('    direct SELECT ... FROM CUSTOMERS -> ',
    Value('SELECT ORIGIN FROM CUSTOMERS'));
  writeln('    SELECT SRC FROM APP.WHICH_ONE    -> ',
    Value('SELECT SRC FROM APP.WHICH_ONE'), '   <- unmoved');
  writeln('    RDB$DEPENDENCIES records         -> ',
    Value('SELECT TRIM(RDB$DEPENDED_ON_SCHEMA_NAME) || ''.'' || TRIM(RDB$DEPENDED_ON_NAME)' +
      ' FROM RDB$DEPENDENCIES WHERE RDB$DEPENDENT_NAME = ''WHICH_ONE'''));

  { -- 5. Plans are schema-qualified. }
  Stmt := A.Prepare(Tr, 'SELECT COUNT(*) FROM CUSTOMERS');
  writeln;
  writeln('plan for unqualified SELECT :', Stmt.GetPlan);

  Tr.Commit;
  writeln;
  writeln('done.');
end.
